### PR TITLE
feat: show no internet dialog when trying to login, sign in (when no net connection present)

### LINKED ISSUE #320: Align text correctly in tickets fragment
Text inside spinner appears a little lower than the rest

### PR DESCRIPTION
**Pull request by dreadpool2**
Fixes #320 #286 
screenshot_20180715-025426
screenshot_20180715-025841

### CODE REVIEW

**Comment by dreadpool2** [comment]
@nikit19 @simarsingh24 @iamareebjamal Please review

**Comment by dreadpool2** [comment]
@iamareebjamal I don't think we can put any logic in VM as if we put the network connection logic, it would involve android classes. I am not able to get which logic should be put.

**Comment by iamareebjamal** [comment]
Check the resource class, that's how you put the logic in VM

**Comment by dreadpool2** [comment]
@iamareebjamal Should I follow this procedure for handling network status in VM? https://developer.android.com/jetpack/docs/guide#addendum

**Comment by iamareebjamal** [comment]
No, look at the resource class in this project

**Comment by dreadpool2** [comment]
@iamareebjamal Thanks a lot! Please review :+1:

File: app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt [hunk 1]
```diff
@@ -29,7 +29,11 @@ class LoginFragment : Fragment() {
             redirectToMain()
 
         rootView.loginButton.setOnClickListener {
-            loginActivityViewModel.login(username.text.toString(), password.text.toString())
+            if (loginActivityViewModel.isNetworkConnected()) {
+                loginActivityViewModel.login(username.text.toString(), password.text.toString())
+            } else {
+                Utils.showNoInternetDialog(activity)
+            }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt]: This if else should be in VM

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt]: PR Updated

File: app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt [hunk 2]
```diff
@@ -29,11 +29,9 @@ class LoginFragment : Fragment() {
             redirectToMain()
 
         rootView.loginButton.setOnClickListener {
-            if (loginActivityViewModel.isNetworkConnected()) {
-                loginActivityViewModel.login(username.text.toString(), password.text.toString())
-            } else {
-                Utils.showNoInternetDialog(activity)
-            }
+            val networkConnected = loginActivityViewModel.isNetworkConnected()
```
[the remark this answers is not in the document]
Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt]: Updated the Pr

**Comment by dreadpool2** [comment]
@iamareebjamal @simarsingh24 Please review

File: app/src/main/java/org/fossasia/openevent/general/data/Resource.kt [hunk 3]
```diff
@@ -13,4 +15,9 @@ class Resource {
 
     fun getString(@StringRes resId: Int, vararg args: Any?) = context.getString(resId, args)
 
+    fun isNetworkConnected(): Boolean {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/data/Resource.kt]: This should be in a separate class called Network. This is not a responsibility of Resource

File: app/src/main/java/org/fossasia/openevent/general/di/Modules.kt [hunk 4]
```diff
@@ -39,6 +40,7 @@ import java.util.concurrent.TimeUnit
 
 val commonModule = applicationContext {
     bean { Preference() }
+    bean { Resource() }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/di/Modules.kt]: AFAIK, this was already listed somewhere

File: app/src/main/java/org/fossasia/openevent/general/auth/SignUpFragment.kt [hunk 5]
```diff
@@ -34,8 +34,8 @@ class SignUpFragment : Fragment() {
             signUp.firstName = firstNameText.text.toString()
             signUp.lastName = lastNameText.text.toString()
             confirmPassword = confirmPasswords.text.toString()
-
             signUpActivityViewModel.signUp(signUp, confirmPassword)
+            if(!signUpActivityViewModel.isNetworkConnected()) Utils.showNoInternetDialog(activity)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/auth/SignUpFragment.kt]: This is not how you do it, subscribe to a livedata which pushes network not available errors. No ifs in activities

File: app/src/main/java/org/fossasia/openevent/general/auth/SignUpFragmentViewModel.kt [hunk 6]
```diff
@@ -78,6 +79,9 @@ class SignUpFragmentViewModel(private val authService: AuthService) : ViewModel(
     }
 
     private fun hasErrors(email: String?, password: String?, confirmPassword: String): Boolean {
+        if (!isNetworkConnected()) {
+            return true
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/auth/SignUpFragmentViewModel.kt]: Push to LiveData about network not available

**Comment by dreadpool2** [comment]
Thanks! Will do the changes :+1:

File: app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt [hunk 7]
```diff
@@ -26,12 +26,12 @@ class LoginFragment : Fragment() {
                               savedInstanceState: Bundle?): View? {
         rootView = inflater.inflate(R.layout.fragment_login, container, false)
 
+        loginActivityViewModel.showNoInternetDialog()
```
[the remark this answers is not in the document]
Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/auth/LoginFragment.kt]: PR updated

File: app/src/main/java/org/fossasia/openevent/general/data/Network.kt [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import org.fossasia.openevent.general.OpenEventGeneral
+
+class Network {
+
+    private val context by lazy {
+        OpenEventGeneral.appContext!!
+    }
+
+    fun isNetworkConnected(): Boolean {
+        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/data/Network.kt]: Make this a lazy property

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/data/Network.kt]: Updated PR

File: app/src/main/java/org/fossasia/openevent/general/data/Network.kt [hunk 9]
```diff
@@ -11,7 +11,7 @@ class Network {
     }
 
     fun isNetworkConnected(): Boolean {
-        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
+        val connectivityManager by lazy { context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/data/Network.kt]: No use unless you make it a global variable

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/data/Network.kt]: Updated

File: app/src/main/java/org/fossasia/openevent/general/utils/Utils.kt [hunk 10]
```diff
@@ -30,4 +31,12 @@ object Utils {
     fun showProgressBar(progressBar: ProgressBar, show: Boolean) {
         progressBar.visibility = if (show) View.VISIBLE else View.GONE
     }
+
+    fun showNoInternetDialog(activity: Activity?) {
+        val builder = AlertDialog.Builder(activity)
+        builder.setMessage(activity?.resources?.getString(R.string.no_internet_message))
+                .setPositiveButton(activity?.resources?.getString(R.string.ok)) { dialog, _ -> dialog.cancel() }
+        val alert = builder.create()
+        alert.show()
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/utils/Utils.kt]: Directly call `alert.show()`

File: app/src/main/java/org/fossasia/openevent/general/utils/Utils.kt [hunk 11]
```diff
@@ -30,4 +31,12 @@ object Utils {
     fun showProgressBar(progressBar: ProgressBar, show: Boolean) {
         progressBar.visibility = if (show) View.VISIBLE else View.GONE
     }
+
+    fun showNoInternetDialog(activity: Activity?) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/utils/Utils.kt]: Take context, not activity

File: app/src/main/java/org/fossasia/openevent/general/data/Network.kt [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package org.fossasia.openevent.general.data
+
+import android.content.Context
+import android.net.ConnectivityManager
+import org.fossasia.openevent.general.OpenEventGeneral
+
+class Network {
+    private val connectivityManager by lazy { context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/data/Network.kt]: Should be below context, and move the body to next line

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/data/Network.kt]: Updated PR :+1:

File: app/src/main/java/org/fossasia/openevent/general/data/Network.kt [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+import android.net.ConnectivityManager
+import org.fossasia.openevent.general.OpenEventGeneral
+
+class Network {
+
+    private val context by lazy {
+        OpenEventGeneral.appContext!!
+    }
+
+    private val connectivityManager by lazy { context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/data/Network.kt]: Move body to next line

Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/data/Network.kt]: Updated PR

**Comment by dreadpool2** [comment]
@iamareebjamal Removed the nullable type too. PR Updated

**Pull request merged**